### DISCUSSION TITLE
Use different templateEngineOverride values for different pages in a pagination set

### DISCUSSION THREAD

**Discussion by zachleat** [11ty/eleventy#4062] [Enhancement Queue]
> What I really want to do is build two things from one source file -- output to manuscript/foo.md (to be consumed by leanpub) and also _site/foo/index.html

Reported by @zcorpan

https://twitter.com/zcorpan/status/1113803186693971970

(There may be a better way to do this, open to ideas)

**Comment by Ryuno-Ki** [comment]
Hm,in shell you would likely use [tee](https://www.gnu.org/software/coreutils/manual/html_node/tee-invocation.html) for that. Here's an approach using writableStreams: https://github.com/jb55/node-tee (albeit needs an update regarding modern expressions)

**Comment by Ryuno-Ki** [comment]
Hey @zachleat
What about this: Instead of turning dat failes + markdown into HTML, turn them into a well-defined JSON.
Then offer hooks to grab that JSON and render it.

This way you could have a single point to render different formats from.

I imagine something along the lines of [what istanbul.js is doing](https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-lib-coverage/lib/file.js) (I recall them having better docs on how to write a custom reporter - or did I confuse it [with karma](https://github.com/karma-runner/karma-junit-reporter/blob/master/index.js)?)

This way, people could come up with weird ideas. Like https://jsonfeed.org/
Or using Eleventy as kinda headless CMS etc.

WDYT?

**Comment by zachleat** [comment]
Hmm @Ryuno-Ki I’m not sure what you mean—can you elaborate? What would the standard JSON format represent?

**Comment by Ryuno-Ki** [comment]
Okay, then let me explain it in more detail.

First off, I'm coming from an MVC background (Ruby on Rails, Django etc).

There, in a view method, you would compute a context object and pass it to a template. This way you render the output format (e.g. HTML or JSON or LaTeX or who-knows).

My mental model was now like this:
Eleventy is currently built up of several parts:
- Markdown files
- Data dirs
- Layout files
- Utilities (filters, shortcodes, plugins etc).

I _assume_ that you compute some kind of context object from the markdown files and data dir. Then passing each of those to the specified layout (either set in frontmatter or data dir - at this point likely doesn't matter).
The resulting HTML is then written to disk.

What about allowing to hook into the computed context object?

It would need to be well defined.
That is, the [JSON schema](https://www.jsonschema.net/) must contain all relevant information (metadata, content, source file, ...).

Then ... renderers could pick those context objects up and treat them like they please (e.g. passing it to a layout file, change the type of the layout template language or so).

It's something I [liked on harp.js](http://harpjs.com/docs/development/rules). Harp.js can even be used as middleware to express.js!

But being able to make network requests at build time + layout hierarchy are niiicee also :-)

**Comment by Ryuno-Ki** [comment]
Maybe would allow things like #488 as well.
E.g. transforming information from other JAMstack/CMS to a single format.

**Comment by Ryuno-Ki** [comment]
Hm, or I am thinking too far ahead. What about working together with [HTMLx](https://github.com/htmlx-org/HTMLx) instead?

You're already supporting quite some template languages...

**Comment by zachleat** [comment]
Just FYI, the next large feature on the roadmap is https://github.com/11ty/eleventy/issues/117 which will decentralize the template language support from Eleventy core, allowing plugins to add their own custom extensions and support template languages.

But let me stew on what you’ve said here a bit more.

**Comment by Ryuno-Ki** [comment]
Yeah, I can see, how this would play into what I envisioned.
Would be a good time to address https://github.com/11ty/eleventy/issues/405#issuecomment-481810206 after its release, I think :-)

**Comment by zachleat** [comment]
> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!
